kv/kvserver: TestRestoreReplicas failed

kv/kvserver.TestRestoreReplicas [failed](https://mesolite.cluster.engflow.com/invocations/default/3f3ab93a-b8e5-4b29-9c1a-cb1f2749cda2?testReportRun=2&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on release-25.2.9-rc @ [f0bfb1cb00838ff45a508e4f1eba087e9835a674](https://github.com/cockroachdb/cockroach/commits/f0bfb1cb00838ff45a508e4f1eba087e9835a674):


```
=== RUN   TestRestoreReplicas
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestRestoreReplicas1064146711
    test_log_scope.go:76: use -show-logs to present logs inline
    client_raft_test.go:412: 
        	Error Trace:	pkg/kv/kvserver/client_raft_test.go:412
        	Error:      	An error is expected but got nil.
        	Test:       	TestRestoreReplicas
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestRestoreReplicas1064146711
--- FAIL: TestRestoreReplicas (20.32s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRestoreReplicas.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-56341

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -392,40 +392,36 @@ func TestRestoreReplicas(t *testing.T) {
 
 	require.NoError(t, tc.Restart())
 
-	// Find the leaseholder and follower. The restart may cause the Raft
-	// leadership to bounce around a bit, since we don't fully enable Raft
-	// prevote, so we loop for a bit until we find the leaseholder.
+	// The restart may cause the Raft leadership and the lease to bounce around a
+	// bit, so we loop until we find the leaseholder and able to propose the
+	// second increment.
 	incArgs = incrementArgs(key, 5)
-	var followerStore *kvserver.Store
 	testutils.SucceedsSoon(t, func() error {
 		var pErr *kvpb.Error
 		for i := 0; i < tc.NumServers(); i++ {
 			_, pErr = kv.SendWrapped(ctx, tc.GetFirstStoreFromServer(t, i).TestSender(), incArgs)
 			if pErr == nil {
-				followerStore = tc.GetFirstStoreFromServer(t, 1-i)
-				break
+				return nil
 			}
 			require.IsType(t, &kvpb.NotLeaseHolderError{}, pErr.GetDetail())
 		}
 		return pErr.GoError()
 	})
 
-	// The follower should now return a NLHE.
-	_, pErr = kv.SendWrapped(ctx, followerStore.TestSender(), incArgs)
-	require.Error(t, pErr.GoError())
-	require.IsType(t, &kvpb.NotLeaseHolderError{}, pErr.GetDetail())
-
-	testutils.SucceedsSoon(t, func() error {
-		getArgs := getArgs(key)
-		if reply, err := kv.SendWrappedWith(ctx, followerStore.TestSender(), kvpb.Header{
-			ReadConsistency: kvpb.INCONSISTENT,
-		}, getArgs); err != nil {
-			return errors.Errorf("failed to read data: %s", err)
-		} else if e, v := int64(28), mustGetInt(reply.(*kvpb.GetResponse).Value); v != e {
-			return errors.Errorf("failed to read correct data: expected %d, got %d", e, v)
-		}
-		return nil
-	})
+	// Both servers should eventually observe the new value.
+	for i := 0; i < tc.NumServers(); i++ {
+		testutils.SucceedsSoon(t, func() error {
+			getArgs := getArgs(key)
+			if reply, err := kv.SendWrappedWith(ctx, tc.GetFirstStoreFromServer(t, i).TestSender(), kvpb.Header{
+				ReadConsistency: kvpb.INCONSISTENT,
+			}, getArgs); err != nil {
+				return errors.Errorf("failed to read data: %s", err)
+			} else if e, v := int64(28), mustGetInt(reply.(*kvpb.GetResponse).Value); v != e {
+				return errors.Errorf("failed to read correct data: expected %d, got %d", e, v)
+			}
+			return nil
+		})
+	}
 
 	validate := func(s *kvserver.Store) {
 		repl := s.LookupReplica(key)

